src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.SpeedOpt.cs

```diff
@@ -244,6 +245,70 @@ private TElement Last(TElement[] items)
         }
     }
 
+    internal sealed partial class OrderedEnumerable<TElement, TKey> : OrderedEnumerable<TElement>
+    {
+        // For complicated cases, rely on the base implementation that's more comprehensive.
+        // For the simple case of OrderBy(...).First() or OrderByDescending(...).First() (i.e. where
```

Comment:
Wouldn't it be even faster if `MinBy` or `MaxBy` were used in this case?

Reply:
> Arguably then it's just an artifact of the particular workaround used to encode MaxBy.

I don't understand. The issue is very specifically that MaxBy uses `>`:
https://github.com/dotnet/runtime/blob/bcc1d3d6f00fbcea3f454b2e35bceeaa51e604b1/src/libraries/System.Linq/src/System/Linq/Max.cs#L488
and the OrderBy.Last semantics need it to be `>=`.

Reply:
Is using `Last` common enough in such cases? grep.app seems to suggest that [`OrderByDescending.First`](https://grep.app/search?q=OrderByDescending%5C%28%5B%5E%5C%29%5D%2A%5C%29.First%5C%28%5C%29&regexp=true&filter[lang][0]=C%23) is more common than [`OrderBy.Last`](https://grep.app/search?q=OrderBy%5C%28%5B%5E%5C%29%5D%2A%5C%29.Last%5C%28%5C%29&regexp=true&filter[lang][0]=C%23) (although the total number of hits don't appear to be large enough for this to be conclusive).

My earlier point was mostly an assertion that in most cases, what chain does get picked is accidental rather than intentional use of the stability guarantees of ordered enumerables, and that in most cases these patterns survive from when `MaxBy` and `MinBy` weren't available -- so perhaps it might be best to guide folks to just use that (e.g. via an analyzer). You're right that these are no substitute for chaining with `FirstOrDefault` though.